stellargraph/data/explorer.py

```diff
@@ -780,3 +783,163 @@ def _check_neighbourhood_sizes(self, in_size, out_size):
+            num_cw (int): Total number of context windows to generate. For comparable
+                results to most other random walks, this should be a multiple of the number
+                of nodes in the graph.
+            cw_size (int): Size of context window. Also used as the minimum walk length,
+                since a walk must generate at least 1 context window for it to be useful.
+            max_walk_length (int): Maximum length of each random walk. Should be greater
+                than or equal to the context window size.
+            initial_edge_bias (str, optional): distribution to use when choosing a random
+                initial temporal edge to start from. Available options are:
+                * None (default) - the initial edge is picked from a uniform distribution
+                * "exponential" - heavily biased towards more recent edges.
+            walk_bias (str, optional): distribution to use when choosing a random
+                neighbour to walk through. Available options are:
+                * None (default) - neighbours are picked from a uniform distribution
+                * "exponential" - exponentially decaying probability, resulting in a bias
+                    towards shorter time gaps
+            p_walk_success_threshold (float): Lower bound for the proportion of successful
+                (i.e. longer than minimum length) walks. If the 95% percentile of the
+                estimated proportion is less than the provided threshold, a RuntimeError
+                will be raised.
+            seed (int, optional): Random number generator seed; default is None
+
+        Returns:
+            List of lists of node ids for each of the random walks
+
+        """
+        if cw_size < 2:
+            raise ValueError(
+                f"cw_size: context window size should be greater than 1, found {cw_size}"
+            )
+        if max_walk_length < cw_size:
+            raise ValueError(
+                f"max_walk_length: maximum walk length should not be less than the context window size, found {max_walk_length}"
+            )
+
+        np_rs = self._np_random_state if seed is None else np.random.RandomState(seed)
+        walks = []
+        num_cw_curr = 0
+
+        edges, times = self.graph.edges(include_edge_weight=True)
+        temporal_edges = list(zip(edges, times))
+        edge_biases = self._temporal_biases(
+            times, None, bias_type=initial_edge_bias, is_forward=False,
+        )
+
+        successes = 0
+        failures = 0
+
+        def not_progressing_enough():
+            # (this uses Beta(1, 1) as the prior, since it's uniform on p)
+            posterior = stats.beta(1 + successes, 1 + failures)
+            return posterior.ppf(0.95) < p_walk_success_threshold
+
+        # loop runs until we have enough context windows in total
+        while num_cw_curr < num_cw:
+            (src, dst), t = self._sample(temporal_edges, edge_biases, np_rs)
```

Comment:
Maybe we could have `_sample` return the index, and then write:

```python
first_edge_index = self._sample(edge_biases, np_rs)
src, dst = edges[first_edge_index]
t = times[first_edge_index]
```

This in particular lets us remove the `temporal_edges = ...` line, where the O(num edges) `zip` with a numpy array is unnecessarily expensive (https://github.com/stellargraph/stellargraph/pull/750#discussion_r374377997).

Reply:
Thanks, although I think I'll still have to do something like this for the `temporal_edges = ...` line for now:
```python
edges, times = self.graph.edges(include_edge_weights=True)
edges = list(edges) # convert networkx EdgeView to list
```
which might still be a bit unnecessarily expensive

I can add a note somewhere to remove that as part of removing the networkx stellargraph somewhere, unless you have other thoughts?

Reply:
Actually I can just use the new stellargraph in the test